Std/Tactic/TryThis.lean

```diff
@@ -1,42 +1,126 @@
+apply the replacement.
+-/
+@[codeActionProvider] def tryThisProvider : CodeActionProvider := fun params snap => do
+  let doc ← readDoc
+  let startPos := doc.meta.text.lspPosToUtf8Pos params.range.start
+  let endPos := doc.meta.text.lspPosToUtf8Pos params.range.end
+  pure <| snap.infoTree.foldInfo (init := #[]) fun _ctx info result => Id.run do
+    let .ofCustomInfo info := info | result
+    let some tti := info.value.get? TryThisInfo | result
+    let some (head, tail) := (tti.span? <|> return (← info.stx.getPos?, ← info.stx.getTailPos?))
+      | result
+    unless head ≤ endPos && startPos ≤ tail do return result
+    result.push {
+      eager.title := "Apply 'Try this'"
+      eager.kind? := "refactor"
+      eager.edit? := WorkspaceEdit.ofTextEdit params.textDocument.uri {
+        range := doc.meta.text.utf8PosToLspRange head tail
+        newText := tti.suggestion
+      }
+    }
+
+/--
+This is a widget which is placed by `TryThis.addSuggestion`; it says `Try this: <replacement>`
+where `<replacement>` is a link which will perform the replacement.
+-/
+@[widget] def tryThisWidget : Widget.UserWidgetDefinition where
+  name := "Tactic replacement"
+  javascript := "
+import * as React from 'react';
+import { EditorContext } from '@leanprover/infoview';
+const e = React.createElement;
+export default function(props) {
+  const editorConnection = React.useContext(EditorContext)
+  function onClick() {
+    editorConnection.api.applyEdit({
+      changes: { [props.pos.uri]: [{ range: props.range, newText: props.suggestion }] }
+    })
+  }
+  return e('div', {className: 'ml1'}, e('pre', {className: 'font-code pre-wrap'}, [
+    'Try this: ',
+    e('a', {onClick, className: 'link pointer dim', title: 'Apply suggestion'}, props.suggestion),
+    props.info
+  ]))
+}"
 
 /-- Replace subexpressions like `?m.1234` with `?_` so it can be copy-pasted. -/
 partial def replaceMVarsByUnderscores [Monad m] [MonadQuotation m]
     (s : Syntax) : m Syntax :=
-  s.replaceM fun s => if let `(?$_:ident) := s then `(?_) else pure none
+  s.replaceM fun s => do
+    let `(?$id:ident) := s | pure none
+    if id.getId.hasNum || id.getId.isInternal then `(?_) else pure none
 
 /-- Delaborate `e` into an expression suitable for use in `refine`. -/
 def delabToRefinableSyntax (e : Expr) : TermElabM Term :=
   return ⟨← replaceMVarsByUnderscores (← delab e)⟩
 
-/-- Add a "try this" suggestion. (TODO: this depends on code action support) -/
-def addSuggestion [Monad m] [MonadLog m] [AddMessageContext m] [MonadOptions m]
-    (origStx : Syntax) (suggestion : Syntax) : m Unit :=
-  logInfoAt origStx m!"{suggestion}"
+/-- Add a "try this" suggestion. -/
+def addSuggestion (origStx : Syntax) {kind : Name} (suggestion : TSyntax kind)
+    (suggestionForMessage : Option MessageData := none)
+    (ref? : Option Syntax := none)
+    (extraMsg : String := "") : MetaM Unit := do
```

Comment:
I think it would be clearer to swap (the names) `origStx` and `ref?` here.  Usually we use `ref` to specify where the error message should go.  While "original syntax" sounds like it should be the original tactic/term that we want to replace using the suggestion.

Reply:
I reorganized the names a bit and added parameter documentation.